source/common/network/connection_impl.cc

```diff
@@ -123,35 +133,29 @@ void ConnectionImpl::close(ConnectionCloseType type) {
     //          ConnectionManagerImpl::checkForDeferredClose()
     //     2) A second close is issued by a subsequent call to
     //        ConnectionManagerImpl::checkForDeferredClose() prior to returning from onData()
-    if (delayed_close_) {
+    if (inDelayedClose()) {
+      // Validate that a delayed close timer is already enabled unless it was disabled via
+      // configuration.
+      ASSERT(!delayed_close_timeout_set || delayed_close_timer_ != nullptr);
+      if (type == ConnectionCloseType::FlushWrite || !delayed_close_timeout_set) {
```

Comment:
Out of curiosity, does this state change actually happen or is this preemptive? Just wondering if we can simplify until this becomes an actual issue? Can we just assert the type is the same?

Reply:
This was just preemptive. I've changed the logic to a stronger ASSERT() and removed unnecessary tests.

Reply:
My latest commit restores this logic. The reason is that after discussing with @alyssawilk offline, I now prefer maintaining `Connection::close()` API backwards compatibility with the existing behavior which allows `ConnectionCloseType` transitions between `close()` calls. This minimizes the (admittedly small) risk that an existing user of the API passing `FlushWrite` and `FlushWriteAndDelay` would break after this PR is merged. It also enforces consistent handling of `type` transitions by allowing all `ConnectionCloseType` transitions as opposed to special casing `FlushWrite` and `FlushWriteAndDelay`.

Another option, which would significantly simplify the `close()` logic, is to enforce that callers use the same `type` after the initial `close()` is issued on a `Connection` but this would break backwards compatibility and would have a much higher risk of breaking existing users.

Reply:
> Another option, which would significantly simplify the close() logic, is to enforce that callers use the same type after the initial close() is issued on a Connection but this would break backwards compatibility and would have a much higher risk of breaking existing users.

This would be my preference. Are there any existing users that actually do this? (There might be and it might a reasonable thing to try a graceful close followed by a force closed, I just don't remember).

Reply:
> This would be my preference. Are there any existing users that actually do this? (There might be and it might a reasonable thing to try a graceful close followed by a force closed, I just don't remember).

Based on a quick scan, I haven't found any filters that attempt a graceful close followed by a forced close (it tends to be one or the other, typically the latter is only used on error handling code paths prior to any other close()s being issued). However, this seems like a reasonable thing to support, and more importantly, the [existing logic](https://github.com/envoyproxy/envoy/blob/master/source/common/network/connection_impl.cc#L80) in the `ConnectionImpl` destructor forces support for `X -> NoFlush` transitions since a "just in case" `close(NoFlush)` is always attempted on destruction. This could be changed, but it seems completely reasonable to me given that leaking connections is much worse than not fully flushing the write buffer or avoiding the race that the `delayed_close_timeout` is mitigating.

So, the modified proposal would be to only allow state transitions `ConnectionCloseType::<any> -> ConnectionCloseType::NoFlush` after the first `close()` has been issued. This still achieves the goal of simplifying the logic while providing some defense against bugs.

Reply:
@alyssawilk @mattklein123 I would prefer to create a second PR for changing the `Connection::close()` API to enforce the state transition constraints. 

This PR should be safe to merge given the amount of review and test coverage added. What do you think about doing this in two steps and I can follow up with the API change after this PR is merged? I think we will end up with a cleaner history and more granular options for rolling back if the API causes breakage.

Reply:
Sure that's fine with me.